accepted/future-releases/records/records-feature-specification.md

```diff
@@ -257,6 +257,80 @@ avoid this ambiguity ([#2469][]).**
 
 [#2469]: https://github.com/dart-lang/language/issues/2469
 
+### Ambiguity with metadata annotations
+
+A metadata annotation may or may not have an argument list following it. A
+variable declaration may omit a preceding type annotation. Likewise, a function
+declaration may omit a preceding return type. This combination of syntax where
+an optional trailing element is followed by syntax with an optional preceding
+element can lead to ambiguity. In particular:
+
+```dart
+@metadata (a, b) function() {}
+```
+
+This could be a metadata annotation `@metadata(a, b)` associated with a function
+declaration with no return type. Or it could be a metadata annotation
+`@metadata` associated with a function whose return type is the record type `(a,
+b)`.
+
+In practice, idiomatically written code is clear thanks to whitespace:
+
+```dart
+@metadata(a, b) function() {}
+
+@metadata (a, b) function() {}
+```
+
+The former applies `(a, b)` to the metadata annotation and the latter is a
+return type. We disambiguate in the same way, by making whitespace after a
+metadata annotation name significant. Change the grammar to:
+
+```
+metadatum ::= identifier                                    // Existing rule.
+            | qualifiedName                                 // Existing rule.
+            | constructorDesignation ~WHITESPACE arguments  // Changed.
+```
+
+The `~WHITESPACE` lexical rule matches when there are no whitespace characters
+(according to the existing `WHITESPACE` lexical rule) between the
+`constructorDesignation` and `arguments`. Comments do not count as whitespace,
+nor do whitespace characters in comments. The newline at the end of a line
+comment *does* count as whitespace.
+
+```dart
+// These are argument lists to the annotation:
+@metadata(x,) a;
+@metadata/* a comment */(x,) a;
+@metadata/* a
+multi
+line
+comment */(x,) a;
+
+// These are record variable types:
+@metadata (x,) a;
+
+@metadata
+(x,) a;
+
+@metadata // Comment.
+(x,) a;
+
+@metadata/* Comment with newline after. */
+(x,) a;
```

Comment:
Other examples that might be relevant
```dart
@metadata /* a */(int. int) Function() f;
@metadata/* a */ (int. int) Function() f;
@metadata/* a *//* b */(int. int) Function() f;  // *
@metadata/* a */ /* b */(int. int) Function() f;
```
Of these, I think only the one marked `*` would be valid.

Again, I think it's just too complicated to allow comments. 

Also, it's worth showing that whitespace after `>` is different:
```dart
@metadata<A> (int, int) Function() f; // *
@metadata<A> 
  (int, int) Function() f;  // *
```
(Both would be valid as *annotation arguments*.)

Reply:
> From https://github.com/dart-lang/language/issues/2469#issuecomment-1255950332
>
>> I'm very tempted to say that if the next source token (ignoring comments and whitespace) after the identifier does not start right where the identifier ends, it's not part of the annotation. So no comments either.
>
> Can we have that instead please?

Yeah, I think that's reasonable. It's almost as if the `(` is actually part of the annotation's identifier *token*.

> Also, it's worth showing that whitespace after `>` is different:
> 
> ```dart
> @metadata<A> (int, int) Function() f; // *
> @metadata<A> 
>   (int, int) Function() f;  // *
> ```
> 
> (Both would be valid as _annotation arguments_.)

Ugh, that's super weird. What I specified in this PR is:

```dart
@metadata<A>(int, int) f; // `(int, int`) goes to annotation.
@metadata <A>(int, int) f; // `(int, int`) goes to annotation.
@metadata<A> (int, int) f; // `(int, int`) goes to variable type.
@metadata <A> (int, int) f; // `(int, int`) goes to variable type.
```

I wonder if we should disallow whitespace before `<` too. It's not ambiguous (now) but it feels very strange to not just ignore the whitespace before `<` but also have that mean that whitespace before the subsequent `(` is now allowed and ignored.

Reply:
OK, I took a position on this. Let me know if you don't like it.